File: python/euler1.py
#! /usr/bin/python3.4
"""Find the sum of all the multiples of two integers below an upper bound. """

from euler import multiple_sum, arith_sum
import time


def euler1():
    print("\nThis program will calculate the sum of the multiples of two specified integers below a given value.")
    mult1 = int(input("\nWhat is the first integer you want to use?\n"))
    mult2 = int(input("\nWhat is the second integer you want to use?\n"))
    upper_bound = int(input("\nPlease specify an upper bound for the calculation:\n"))

    # -- This was the original version and it is FASTER than a list comp...
    print('--> Take 1: Original solution of multiple function calls and calculating THREE sums (m,n and overlap)...')
    s1 = time.time()
    overlap = mult1 * mult2
    upper_bound = upper_bound-1
    # corrects the upper bound so that the for loop will calculate BELOW that value rather than up to it
    grand_total = multiple_sum(mult1, upper_bound) + multiple_sum(mult2, upper_bound) - multiple_sum(overlap, upper_bound)
    f1 = time.time() - s1
    print(">>> The grand total is %d" % (grand_total))
    print("Solution found in " + str("{0:.5f}".format(f1)) + " seconds.\n")

    print('--> Take 2: Single for loop checking divison modulo m/n to add to a running total...')
    s2 = time.time()
    total = 0
    for n in range(1, upper_bound+1):
        if n % mult1 == 0 or n % mult2 == 0:
            total += n
    f2 = time.time() - s2
    print(">>> The grand total is %d" % (total))
    print("Solution found in " + str("{0:.5f}".format(f2)) + " seconds.\n")


    # -- This is a one liner but the list comp is SLOWER than the multiple function calls!
    print('--> Take 3: One liner with a list comprehension...')
    s3 = time.time()
    grand_total = sum([x for x in range(1, upper_bound+1) if x % mult1 == 0 or x % mult2 == 0])
    f3 = time.time() - s3
    print(">>> The grand total is %d" % (grand_total))
    print("Solution found in " + str("{0:.5f}".format(f3)) + " seconds.\n")


    print('--> Take 4: Using a formula for the sum of an arithmetic series...')
    s4 = time.time()
    n1 = upper_bound//mult1
    n2 = upper_bound//mult2
    n3 = upper_bound//overlap
    total = arith_sum(n1, mult1, mult1) + arith_sum(n2, mult2, mult2) - arith_sum(n3, overlap, overlap)
    f4 = time.time() - s4
    print(">>> The grand total is %d" % (total))
    print("Solution found in " + str("{0:.5f}".format(f4)) + " seconds.\n")


if __name__ == "__main__":
    euler1()

File: go/euler1.go
// Find the sum of all the multiples of 3 or 5 below 1000.
package main

import "fmt"

func multiple_sum(base, upper_bound int) int {
	var total int = 0
	roof := int(upper_bound / base)
	fmt.Printf("\n    There are %d %ds below %d", roof, base, (upper_bound + 1))
	for i := 0; i < roof; i++ {
		total += base * (int(i) + 1.0)
	}
	fmt.Printf("\n    The sum of the multiples of %d = %d", base, total)
	return total
}

func aSum(A, D, N int) int {
	/*	Floats cause problems!!!
		-> need to use float64 for precision but there is no
		warning of overflow unless I manually check!
		Also, setting the upper bound too high makes the result negative...
		-> Probably a two's compliment problem? */
	a := float64(A)
	d := float64(D)
	n := float64(N)
	Sn := int(n * (2.0*a + (n-1.0)*d) / 2.0)
	return int(Sn)
}

func main() {
	fmt.Println("This program will calculate the sum of the multiples of two specified integers below a given value.")
	var (
		mult1       int
		mult2       int
		upper_bound int
	)
	fmt.Println("What is the first integer you want to use?")
	fmt.Scanf("%d", &mult1)
	fmt.Println("What is the second integer you want to use?")
	fmt.Scanf("%d", &mult2)
	overlap := mult1 * mult2
	fmt.Println("Please specify an upper bound for the calculation:")
	fmt.Scanf("%d", &upper_bound)
	fmt.Printf("\n  Calculating the sum of the multiples of %d and %d below %d", mult1, mult2, upper_bound)
	upper_bound -= 1

	fmt.Println("\n--> Take 1: using a for loop to sum up required values and subtract the overlap:")
	grand_total := multiple_sum(mult1, upper_bound) + multiple_sum(mult2, upper_bound) - multiple_sum(overlap, upper_bound)
	fmt.Println("\n\nThe grand total is: ", grand_total)

	fmt.Println("\n--> Take 2: using the sun of an arithmetic series:")
	n1 := upper_bound / mult1
	n2 := upper_bound / mult2
	n3 := upper_bound / (mult1 * mult2)
	S1 := aSum(mult1, mult1, n1)
	S2 := aSum(mult2, mult2, n2)
	S3 := aSum(overlap, overlap, n3)
	fmt.Println(mult1, n1, S1)
	fmt.Println(mult2, n2, S2)
	fmt.Println(overlap, n3, S3)
	total := S1 + S2 - S3
	fmt.Println("\n\nThe grand total is: ", total)
}
